### ISSUE TITLE
Using Pysaml2 with Fargate

### ISSUE THREAD

**Issue by JohnnyP10** [IdentityPython/pysaml2#572]
It seems when running pysaml2 in a Fargate Container the outstanding_queries are never populate which results in an Unsolicited Response.  The only solution i have found is to add the following to the SP config 'allow_unsolicited': True
## Code Version

the latest as of 2018_11_27 which is 4.6.4
## Expected Behavior

It should retain the outstanding queries
## Current Behavior

the struct is empty
## Possible Solution

I saw mention of State in the client.py but nothing on how to use it.  Maybe this would be a solution?
## Steps to Reproduce

1. Configure a python app with default values and try running it in fargate.
2.
3.
4.

**Comment by c00kiemon5ter** [comment]
Hello, this [has come up before](https://github.com/IdentityPython/pysaml2/issues/333#issuecomment-419934255) and is unrelated to the environment that pysaml2 runs in (in this case _fargate_).

`outstanding_queries` is just a dictionary structure - a mapping between session-IDs and `Request` objects. In order to decouple the persistence mechanism that caller want to use from the library itself, it is the caller's responsibility to actually populate those entries (write them to the long-term storage medium; a database, a file, in memory). The library does make the appropriate checks and raises an `UnsolicitedResponse` exception when such a response is received. Again, it is the callers responsibility to act on whether such an exception was raised or not and remove items from the structure following some policy.
In other words, `outstanding_queries` needs to be backed with persistent storage. The caller is responsible to keep the storage and the structure in sync, by adding and removing entries on the storage medium and the structure as needed, in parallel.

_I am open to re-thinking this; whether it should be restructured and how_, but certainly it is not a priority at the moment.

**Comment by JohnnyP10** [comment]
@c00kiemon5ter 

Thank you for your response.  That makes sense and has helped me try to debug it.  I am using Django which seems to store this information in Session.  What i have found is that Fargate issues a session when you go to the app but then a second one after you login which doesn't have any of the session info from the first one and hence the Unsolicited Request.  This doesn't happen locally in Docker.  I am contacting AWS to see if they have any ideas why this would be the case.

**Comment by tyctor** [comment]
i am encounteering same problem on AWS fargate, may i ask what was your solution?

**Comment by JohnnyP10** [comment]
In my SP definition i had to set 'allow_unsolicited' to True.

**Comment by tyctor** [comment]
thanks, it works
this is because multiple containers can serve single authentication workflow (because of loadbalancer in front), right?

**Comment by JohnnyP10** [comment]
i think as it was pointed out above it needs some sort of data persistence layer.

**Comment by tyctor** [comment]
yes, i reread it, it make sense
in django it can be easily done via signed cookie
